radixdlt-java/src/main/java/com/radixdlt/client/application/translate/StatefulActionToParticlesMapper.java

```diff
@@ -0,0 +1,64 @@
+package com.radixdlt.client.application.translate;
+
+import com.radixdlt.client.atommodel.accounts.RadixAddress;
+import com.radixdlt.client.core.atoms.particles.SpunParticle;
+import io.reactivex.Observable;
+import java.util.Objects;
+
+/**
+ * Maps a high level application action to lower level spun particles used
+ * to construct an atom given a context requirement which this interface describes
+ * via requiredState().
+ */
+public interface StatefulActionToParticlesMapper {
+	class RequiredShardState {
+		private final Class<? extends ApplicationState> stateClass;
+		private final RadixAddress address;
+
+		public RequiredShardState(Class<? extends ApplicationState> stateClass, RadixAddress address) {
+			Objects.requireNonNull(stateClass);
+			Objects.requireNonNull(address);
+
+			this.stateClass = stateClass;
+			this.address = address;
+		}
+
+		/**
+		 * Retrieves the type of application state needed for this requirement
+		 *
+		 * @return the type of application state
+		 */
+		public Class<? extends ApplicationState> stateClass() {
+			return stateClass;
+		}
+
+		/**
+		 * Retrieves the shardable address which needs to be queried to construct the application state
+		 *
+		 * @return the shardable address
+		 */
+		public RadixAddress address() {
+			return address;
+		}
+	}
+
+	/**
+	 * Retrieves the necessary application state to be used in creating new particles
+	 * given a high level action. The returned Observable describes the shardable and
+	 * the type of state needed to create particles.
+	 *
+	 * @param action the action to get the required context about
+	 * @return observable of required contexts required to create spun particles for the action
+	 */
+	Observable<RequiredShardState> requiredState(Action action);
```

Comment:
Why does this need to be an `Observable`? Is there any async logic here? Would a simple `Collection` suffice?

Reply:
Oh no. The turtles..